### ISSUE TITLE
inconsistent data

### ISSUE THREAD

**Issue by algoritmsystems** [shayonj/pg_easy_replicate#167]
During replication i create a test table in the source db:
> SELECT * FROM test_table;
 id |     name      |         created_at         
----+---------------+----------------------------
  1 | Test Record 1 | 2024-10-02 06:00:01.712611
  2 | Test Record 2 | 2024-10-02 06:00:01.712611
  3 | Test Record 3 | 2024-10-02 06:00:01.712611
  4 | Test Record 4 | 2024-10-02 07:38:50.613114
(4 rows)

but after sync in target db data is not consistent:
>SELECT * FROM test_table;
 id | name | created_at 
----+------+------------
(0 rows)

Do I need to create all the records before starting the sync?
Thanks!

**Comment by shayonj** [comment]
Hi!

Thanks for report. Can you share what your bootstrap, start_sync commands are like ?

As long as the tables are part of the publication the data should continue streaming. You can also call the `stats` command to see the status of the replication. Feel free to share that here and happy to help debug

**Comment by shayonj** [comment]
> During replication i create a test table in the source db:

I wonder if the table is actually not part of the publication, i.e it wasn't present when `start_sync` was called - it sounds like thats the case?

**Comment by algoritmsystems** [comment]
> Thanks for report. Can you share what your bootstrap, start_sync commands are like ?
> 
Hi @shayonj !
Thank you for your response!
> pg_easy_replicate bootstrap --group-name database-cluster-dmed --copy-schema
> pg_easy_replicate start_sync --group-name database-cluster-dmed --track-ddl

> As long as the tables are part of the publication the data should continue streaming. You can also call the `stats` command to see the status of the replication. Feel free to share that here and happy to help debug
> pg_easy_replicate stats --group-name database-cluster-dmed
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.376+05:00","v":0,"msg":"(0.000449s) SET standard_conforming_strings = ON","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.376+05:00","v":0,"msg":"(0.000447s) SET client_min_messages = 'WARNING'","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.377+05:00","v":0,"msg":"(0.000285s) SET DateStyle = 'ISO'","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":20,"time":"2024-10-02T15:22:18.377+05:00","v":0,"msg":"Connection established","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.377+05:00","v":0,"msg":"(0.000317s) BEGIN","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.378+05:00","v":0,"msg":"(0.000855s) SET statement_timeout to '5s'","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.381+05:00","v":0,"msg":"(0.002079s)   SELECT\n  s.subname AS subscription_name,\n  c.relnamespace :: regnamespace :: text as table_schema,\n  c.relname as table_name,\n  rel.srsubstate as replication_state\nFROM\n  pg_catalog.pg_subscription s\n  JOIN pg_catalog.pg_subscription_rel rel ON rel.srsubid = s.oid\n  JOIN pg_catalog.pg_class c on c.oid = rel.srrelid\nWHERE s.subname = 'pger_subscription_database_cluster_dmed'\n","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.381+05:00","v":0,"msg":"(0.000326s) COMMIT","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.400+05:00","v":0,"msg":"(0.000820s) SET standard_conforming_strings = ON","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.400+05:00","v":0,"msg":"(0.000450s) SET client_min_messages = 'WARNING'","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.401+05:00","v":0,"msg":"(0.000400s) SET search_path = \"pger\"","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.401+05:00","v":0,"msg":"(0.000381s) SET DateStyle = 'ISO'","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":20,"time":"2024-10-02T15:22:18.402+05:00","v":0,"msg":"Connection established","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.404+05:00","v":0,"msg":"(0.002050s) SELECT CAST(current_setting('server_version_num') AS integer) AS v","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.415+05:00","v":0,"msg":"(0.010523s) SELECT * FROM \"groups\" WHERE (\"name\" = 'database-cluster-dmed') LIMIT 1","version":"0.3.2"}
{"name":"pg_easy_replicate","hostname":"ansible56","pid":4787,"level":30,"time":"2024-10-02T15:22:18.434+05:00","v":0,"msg":"(0.001154s) SET standard_conforming_strings = ON","version":"0.3.2"}
...
{
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "news_categories",
      "replication_state": "data_is_being_copied"
    },
    {
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "reimbursement_drugs",
      "replication_state": "data_is_being_copied"
    },
    {
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "reimbursement_category_disease_codes",
      "replication_state": "data_is_being_copied"
    },
    {
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "references_groups",
      "replication_state": "data_is_being_copied"
    },
    {
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "drug_producers",
      "replication_state": "data_is_being_copied"
    }
  ],
  "replication_stats_count_by_state": {
    "data_is_being_copied": 256,
    "replicating": 2
  },
  "message_lsn_receipts": [
    {
      "received_lsn": "1CF/2E7A4FA0",
      "last_msg_send_time": "2024-10-02 10:22:16 UTC",
      "last_msg_receipt_time": "2024-10-02 10:22:14 UTC",
      "latest_end_lsn": "1CF/2E7A4FA0",
      "latest_end_time": "2024-10-02 10:22:16 UTC"
    },
    {
      "received_lsn": null,
      "last_msg_send_time": "2024-10-02 10:22:17 UTC",
      "last_msg_receipt_time": "2024-10-02 10:22:17 UTC",
      "latest_end_lsn": null,
      "latest_end_time": "2024-10-02 10:22:17 UTC"
    }
  ],
  "sync_started_at": "2024-10-02 06:49:09 UTC",
  "sync_failed_at": null,
  "switchover_completed_at": null
}

**Comment by algoritmsystems** [comment]
> If so, then you can setup a new `bootstrap` and `start_sync` with a new group name (`-g`) for that table and it should work

I need to synchronouslu restore with minimum RTO and RPO from existing DB where thousens of tables ...

**Comment by shayonj** [comment]
Ah that is it, as you can see from the stats, you have 256 tables are still being copied `data_is_being_copied` and only two have reached the replicating state, which is after the initial sync is fully complete. 

```
"replication_stats_count_by_state": {
"data_is_being_copied": 256,
"replicating": 2
}
```

So as long as `test_table` was present in the database when you started the sync, assuming you opted to start sync for all tables, once the initial copy is completed the new data will show up as well. 

But note - if you choose to start sync for all tables, those tables need to be present in the DB at the time of start sync. 

Hope that helps :).

**Comment by shayonj** [comment]
I am closing, but feel free to reopen if you have any further questions :)

**Comment by algoritmsystems** [comment]
just one question: is it necessary to run the command for switchover after start_sync?
>pg_easy_replicate switchover --group-name database-cluster-dmed

or it isn't needed?
Thanks!

**Comment by shayonj** [comment]
> pg_easy_replicate switchover --group-name database-cluster-dmed


Switchover will put the source database in read only and update sequence in target database, plus few other things. So it depends on your use case, if you want continuous replication, then you likely don't need it, but if you have a blue/green setup, then you do. 

> pg_easy_replicate start_sync --group-name database-cluster-dmed --track-ddl --jobs 6


There is no `--job` on start_sync, since pg_easy_replicate just helps with bootstrapping. For more concurrent replication, you can tweak `[max_logical_replication_workers](https://www.postgresql.org/docs/current/runtime-config-replication.html#GUC-MAX-LOGICAL-REPLICATION-WORKERS)` and [max_worker_processes](https://www.postgresql.org/docs/current/runtime-config-resource.html#GUC-MAX-WORKER-PROCESSES) on the subscriber side. But, you will need to do that before starting replication

**Comment by algoritmsystems** [comment]
Hi @shayonj !
How do I know the succesfull end of sychronization?
>pg_easy_replicate stats --group-name database-cluster-dmed
...
    {
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "working_times",
      "replication_state": "replicating"
    },
    {
      "subscription_name": "pger_subscription_database_cluster_dmed",
      "table_schema": "public",
      "table_name": "group_service_analysis",
      "replication_state": "replicating"
    }
  ],
  "replication_stats_count_by_state": {
    "replicating": 257
  },
  "message_lsn_receipts": [
    {
      "received_lsn": "1D7/1A084778",
      "last_msg_send_time": "2024-10-03 06:53:34 UTC",
      "last_msg_receipt_time": "2024-10-03 06:53:33 UTC",
      "latest_end_lsn": "1D7/1A084778",
      "latest_end_time": "2024-10-03 06:53:34 UTC"
    }
  ],
  "sync_started_at": "2024-10-03 01:17:31 UTC",
  "sync_failed_at": null,
  "switchover_completed_at": null
}

From this output, can we know that the recovery was successful?

Is there any need to send a final command, after this sync command?
>pg_easy_replicate start_sync --group-name database-cluster-dmed --track-ddl

Thanks!

**Comment by shayonj** [comment]
> How do I know the succesfull end of sychronization?



With logical replication there is really no "end", because it is always streaming. The call/ask is on the user to decide what needs to happen once all tables are replicating. If you want it to be always replicating, you can leave it as. If you want to move/migrate over to the new DB, you can call the `switchover`, this will make the user on source database in read only mode and update sequences + run analyze in the new DB, so it is ready to accept writes.

**Comment by algoritmsystems** [comment]
> switchover, this will make the user on source database in read only mode and update sequences + run analyze in the new DB, so it is ready to accept writes.

how to restore source db from read only mode to read/write mode after switchover?
